### ISSUE TITLE
Crash on nested _mm_hadd_epi32

### ISSUE THREAD

**Issue by j-jorge** [llvm/llvm-project#53247]
Compiling this program with `clang++ -O1 -mavx2 -c ./hadd_avx2_fail.cpp -o /dev/null` causes the compiler to crash:

```c++
#include <immintrin.h>

int foo()
{
    const __m128i sum = _mm_setzero_si128();

    const __m128i sum1 = _mm_hadd_epi32(sum, sum);
    const __m128i sum2 = _mm_hadd_epi32(sum1, sum);

    return _mm_cvtsi128_si32(sum2);
}
```
The message before the crash is "Overran sorted position", emitted by X86 DAG->DAG Instruction Selection.

Reduced IR code looks like this:
```llvm
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-linux-gnu"

; Function Attrs: mustprogress nofree nosync nounwind readnone uwtable willreturn
define dso_local i32 @foo() local_unnamed_addr #0 {
entry:
  %0 = call <4 x i32> @llvm.x86.ssse3.phadd.d.128(<4 x i32> zeroinitializer, <4 x i32> zeroinitializer) #3
  %1 = call <4 x i32> @llvm.x86.ssse3.phadd.d.128(<4 x i32> %0, <4 x i32> zeroinitializer) #3
  %vecext.i = extractelement <4 x i32> %1, i32 0
  ret i32 %vecext.i
}

; Function Attrs: nofree nosync nounwind readnone
declare <4 x i32> @llvm.x86.ssse3.phadd.d.128(<4 x i32>, <4 x i32>) #1

attributes #0 = { mustprogress nofree nosync nounwind readnone uwtable willreturn "frame-pointer"="none" "min-legal-vector-width"="128" "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-cpu"="x86-64" "target-features"="+avx,+avx2,+cx8,+fxsr,+mmx,+popcnt,+sse,+sse2,+sse3,+sse4.1,+sse4.2,+ssse3,+x87,+xsave" "tune-cpu"="generic" }
attributes #1 = { nofree nosync nounwind readnone }
attributes #2 = { nofree nosync nounwind readnone speculatable willreturn }
attributes #3 = { nounwind }
```

See the [attached archive](https://github.com/llvm/llvm-project/files/7883089/hadd_avx2_fail.zip) for original source, logs, and crash reproducer script.

Tested with clang compiled from commit fc043d8a256b1e431aa1297dffba154a7dd46b25.

**Comment by RKSimon** [comment]
Confirmed the crash in fc043d8a256b1e431aa1297dffba154a7dd46b25 but it seems to be working in trunk - I'll see if I can bisect

**Comment by RKSimon** [comment]
@j-jorge Are you just seeing this on the 13.x branch?

**Comment by j-jorge** [comment]
@RKSimon I did not try many configurations. I have the crash with:
 - clang built from branch release/13.x,
 - clang 13 from Ubuntu repositories,
 - icx 2022, a.k.a Intel's compiler based on Clang (trunk-based, it seems, as the tools report a version 14).

**Comment by j-jorge** [comment]
Branch `release/12.x,` (fed41342a82f5a3a9201819a82bf7a48313e296b) does not have the issue.

**Comment by RKSimon** [comment]
git bisect has finally finished - it looks like I fixed this in 468ff703e114599ce8fb7457bd3c7ef0b219e952 (Issue #51316)

**Comment by RKSimon** [comment]
@tstellar I haven't been keeping track of 13.x for a little while - would it be feasible to cherry pick 468ff703e114599ce8fb7457bd3c7ef0b219e952 do you think?

**Comment by tstellar** [comment]
Merged: 1f1854ce0fe5 75e33f71c2da

**Comment by RKSimon** [comment]
Thanks @tstellar
